Bluetooth: host: Ignore reserved SMP PDUs

jori-nordic opened the pull request:
Do not abort pairing when peer sends a SMP command with an opcode
reserved for future use, as per spec.

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>

jori-nordic committed:
Bluetooth: host: Ignore reserved SMP PDUs

Do not abort pairing when peer sends a SMP command with an opcode
reserved for future use, as per spec.

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>

jhedberg reviewed (approved):
Looks fine to me

jhedberg commented:
Btw, I guess the only difference here is the different log message, since reserved opcodes would already be caught by the `hdr->code >= ARRAY_SIZE(handlers)` check blow

jori-nordic commented:
not really, the check below aborts the pairing (smp_error), which is why I have raised this PR

joerchan commented on subsys/bluetooth/host/smp.c:
```diff
@@ -4266,6 +4266,16 @@ static int bt_smp_recv(struct bt_l2cap_chan *chan, struct net_buf *buf)
+	 * If a packet is received with a Code that is reserved for future use
+	 * it shall be ignored.
+	 */
+	if (hdr->code >= BT_SMP_CMD_RESERVED) {
```
This looks like it is handling parts of the if-statement below, so that could be updated.
In fact we could just re-use that check and not use the CMD_RESERVED, that would make it easier to add new handlers as we don't have to adjust the CMD_RESERVED.

`if (hdr->code >= ARRAY_SIZE(handlers))`

jori-nordic replied:
I did not do that because of the keypress notification: we dont set the keypress bit, and then we abort pairing upon receiving one since no handler is defined. I don't know how zephyr wants to handle this, keep as-is or ignore. If we ignore it I can modify the check below and remove this one

joerchan replied:
Thanks for explaining, I did not notice the missing 0xe in between there.
I checked it up in the specification and it says that it shall be ignored.

```
The keypress field is a 1-bit flag that is used only in the Passkey Entry
protocol and shall be ignored in other protocols. When both sides set that
field to one ....
```

Although I'd prefer to ignore it more explicitly, instead of as an addition to the reserved SMP codes.
So I would prefer to add the handler for the missing SMP keypress PDUs  which simply ignores them
And do that in a commit preceding this commit, that would be cleanest IMO.

jhedberg commented:
> not really, the check below aborts the pairing (smp_error), which is why I have raised this PR

Right, I missed the `smp_error()` difference.

jori-nordic committed:
Bluetooth: host: Add keypress notification SMP handler

Add a dummy SMP command handler for keypress notification, that does
nothing (yet). This allows the next commit to work properly.

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>

jori-nordic commented:
I still have to update the PR, during testing (5mins ago) I found I had to actually allow the keypress command to be received, adding the handler is not enough. I will update later today
